dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp

```diff
@@ -246,136 +172,292 @@ bool tryResetPeerIdInRegion(RegionPtr region, const RegionLocalState & region_st
+    GET_METRIC(tiflash_fap_task_state, type_writing_stage).Increment();
+    SCOPE_EXIT({ GET_METRIC(tiflash_fap_task_state, type_writing_stage).Decrement(); });
+
+    CheckpointInfoPtr checkpoint_info;
+    RegionPtr region;
+    RaftApplyState apply_state;
+    RegionLocalState region_state;
+    std::tie(checkpoint_info, region, apply_state, region_state) = checkpoint;
 
-        Stopwatch watch_ingest;
-        auto kvstore = server->tmt->getKVStore();
-        kvstore->handleIngestCheckpoint(region, checkpoint_info, *server->tmt);
+    auto & storages = tmt.getStorages();
+    auto keyspace_id = region->getKeyspaceID();
+    auto table_id = region->getMappedTableID();
+    auto storage = storages.get(keyspace_id, table_id);
+    if (!storage)
+    {
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Can't get storage engine keyspace_id={} table_id={}",
+            keyspace_id,
+            table_id);
+    }
+    if (storage->engineType() == TiDB::StorageEngine::DT)
+    {
+        auto dm_storage = std::dynamic_pointer_cast<StorageDeltaMerge>(storage);
+        auto new_key_range = DM::RowKeyRange::fromRegionRange(
+            region->getRange(),
+            table_id,
+            storage->isCommonHandle(),
+            storage->getRowKeyColumnSize());
 
-        // Write raft log to uni ps
-        UniversalWriteBatch wb;
-        RaftDataReader raft_data_reader(*(checkpoint_info->temp_ps));
-        raft_data_reader.traverseRemoteRaftLogForRegion(
+        auto segments = dm_storage->buildSegmentsFromCheckpointInfo(new_key_range, checkpoint_info, settings);
+        fap_ctx->insertCheckpointIngestInfo(
+            tmt,
             region_id,
-            [&](const UniversalPageId & page_id, PageSize size, const PS::V3::CheckpointLocation & location) {
-                LOG_DEBUG(
-                    log,
-                    "Write raft log size {}, region_id={} index={}",
-                    size,
-                    region_id,
-                    UniversalPageIdFormat::getU64ID(page_id));
-                wb.putRemotePage(page_id, 0, size, location, {});
-            });
-        auto wn_ps = server->tmt->getContext().getWriteNodePageStorage();
-        wn_ps->write(std::move(wb));
-        GET_METRIC(tiflash_fap_task_duration_seconds, type_ingest_stage).Observe(watch_ingest.elapsedSeconds());
-        GET_METRIC(tiflash_fap_task_result, type_succeed).Increment();
+            new_peer_id,
+            checkpoint_info->remote_store_id,
+            region,
+            std::move(segments),
+            start_time);
+    }
+    else
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "ingest into unsupported storage engine");
+    }
 
-        return genFastAddPeerRes(
-            FastAddPeerStatus::Ok,
-            apply_state.SerializeAsString(),
-            region_state.region().SerializeAsString());
+    // Write raft log to uni ps, we do this here because we store raft log seperately.
```

Comment:
We ingest raft log here, because it is the first peer of this region in this store. And a legacy snapshot will do clean first.